fix for #10761 ; expand wildcard classpath entries

philwalk opened the pull request:
This fixes #10761

Wildcard classpath entries are expanded, which solves 2 problems:

1. The `-save` option requires expanded classpath wildcard entries in jar manifest
2. The runtime classpath requires expanding `-classpath` wildcard entries

The solution is not os-dependent, the jar manifest `Class-Path` property requires expanded wildcards.

The callback `classpath:String` parameter  in [compiler/src/dotty/tools/scripting/ScriptingDriver.scala](compiler/src/dotty/tools/scripting/ScriptingDriver.scala) was changed to `classpathEntries:Seq[String]` so the expanded classpath could also be seen by [compiler/src/dotty/tools/scripting/Main.scala](compiler/src/dotty/tools/scripting/Main.scala) and could to be written to the jar manifest.

Two similar use-cases fixed by the PR each required a different fix:

A classpath consisting of a single wildcard entry, specified on the command line (fixed by appending `$PSEP` to the classpath):
```bash
$ scala -classpath 'lib/*' script.sc
```
A classpath consisting of a single wildcard entry, specified in a hashbang (fixed by accepting '-classpath' and 'lib/*' passed together as a single argument to `dist/bin/scala`).  An example of this use-case:
```scala
#!/opt/scala3/bin/scala -classpath 'lib/*'
def main(args: Array[String]): Unit =
  val psep = sys.props("path.separator")
  for entry <- sys.props("java.class.path").split(psep) do
    printf("%s\n",entry)
```

michelou commented:
I'm wondering if the Scripting execution engine is the only place in the Scala 3 compiler that has to take care of wildcard classpaths. Any thoughts about it ?!

philwalk commented:
I'm not aware of where else in the Scala 3 compiler there might also be a need to expand classpath.  I wasn't able to produce problems except those that this PR addresses.

smarter commented:
> I'm not aware of where else in the Scala 3 compiler there might also be a need to expand classpath

My comment in https://github.com/lampepfl/dotty/issues/10761#issuecomment-743478749 points you to the code in the compiler which currently handles expanding wildcard, reusing that code would be good.

philwalk commented:
@smarter - thanks, I updated the PR.


[compiler/src/dotty/tools/scripting/ScriptingDriver.scala](https://github.com/lampepfl/dotty/blob/f5daaa2cdbde14d5752be9485d2e72db1720049b/compiler/src/dotty/tools/scripting/ScriptingDriver.scala)  now uses the `expandPath()` method in [compiler/src/dotty/tools/io/ClassPath.scala](https://github.com/lampepfl/dotty/blob/f5daaa2cdbde14d5752be9485d2e72db1720049b/compiler/src/dotty/tools/io/ClassPath.scala).

`expandPath()` was extended to include classpath entries with forward slash, regardless of OS.

Test class added to verify:
-  multiple wildcard entries in the classpath
-  bogus classpath entries don't cause a crash

philwalk commented:
@michelou 
The existing `wildSuffix` test assumes that Windows classpaths use only backslash, but the jdk supports forward slash as well.
The classpath reported in #10761  `-classpath 'c:/opt/uejlib2.13/*'`, has forward slash.

The additional test will short circuit in Linux, and is never performed, but in Windows, it prevents the expansion from failing.

An alternative might be something like this:
```scala
  /** Expand single path entry */
  private def expandS(_pattern: String): List[String] = {
    val wildSuffix = "/*"
    val pattern = _pattern.replace('\\','/')

    /* Get all subdirectories, jars, zips out of a directory. */
    def lsDir(dir: Directory, filt: String => Boolean = _ => true) =
      dir.list.filter(x => filt(x.name) && (x.isDirectory || isJarOrZip(x))).map(_.path).toList

    if (pattern == "*") lsDir(Directory("."))
    else if (pattern.endsWith(wildSuffix)) lsDir(Directory(pattern dropRight 2))
    else if (pattern.contains('*')) {
      try {
        val regexp = ("^" + pattern.replace("""\*""", """.*""") + "$").r
        lsDir(Directory(pattern).parent, regexp.findFirstIn(_).isDefined)
      }
      catch { case _: PatternSyntaxException => List(pattern) }
    }
    else List(pattern)
  }
```

philwalk commented:
@michelou - comment was added to line 135 of **[compiler/src/dotty/tools/io/ClassPath.scala](https://github.com/lampepfl/dotty/blob/15f9965316c466df08929b237b9fb0aee9fca52d/compiler/src/dotty/tools/io/ClassPath.scala)** as per your suggestion

philwalk commented:
After the change to leverage `ClassPath#expandPath`, a stack dump similar to the  original error reported in #10761 has recurred in the `-classpath lib/*` use case.

I'm investigating,

philwalk commented:
Added code to expand wildcards in [compiler/src/dotty/tools/dotc/core/MacroClassLoader.scala](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/src/dotty/tools/dotc/core/MacroClassLoader.scala).

I will improve the coverage in the classpath test [compiler/test/dotty/tools/io/ClasspathTest.scala](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/test/dotty/tools/io/ClasspathTest.scala)

philwalk commented:
It turns out that Windows jdk 11 (and maybe earlier versions too?) expand all command line arguments that look like wildcard classpath entries. 

As a result, -classpath args on the command line are expanded (Windows only), even if escaped and quoted, etc..

If a classpath is specified in the hashbang header,  it doesn't get expanded.  The most recent commit covers this use case.

michelou commented:
@philwalk In file [`ClasspathTest.scala`](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/test/dotty/tools/io/ClasspathTest.scala) duplicate import on [line 22](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/test/dotty/tools/io/ClasspathTest.scala#L22).

philwalk commented:
@michelou 




> @philwalk In file [`ClasspathTest.scala`](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/test/dotty/tools/io/ClasspathTest.scala) duplicate import on [line 22](https://github.com/lampepfl/dotty/blob/c58106c8b43ca322879a231df504d1751affd625/compiler/test/dotty/tools/io/ClasspathTest.scala#L22).

I deleted the extra import, and will commit after I add some more tests (hopefully soon).

philwalk commented:
While testing this PR, I discovered an unrelated problem with [dist/bin/scalac](https://github.com/lampepfl/dotty/blob/4bfa8024b712ad029c368661d2064084e5a49a04/dist/bin/scalac)
Here's the simple fix:
```diff
         -repl) PROG_NAME="$ReplMain" && shift ;;
-      -script) PROG_NAME="$ScriptingMain" && target_script="$2" && in_scripting_args=true && shift && shift ;;
+      -script) PROG_NAME="$ScriptingMain" && target_script="$2" && in_scripting_args=true && shift && shift
+               while [[ $# -gt 0 ]]; do addScripting "$1" && shift ; done ;;
      -compile) PROG_NAME="$CompilerMain" && shift ;;
```
The bug manifests itself if the `target_script` also has a `-script` argument.

What's the recommended process for submitting a fix?
Should I file a bug report first?  
Should I add the fix to this PR, or create a new PR?

smarter commented:
Since this PR is complicated enough as it is, I suggest making another PR for that unrelated change.

liufengyun commented:
@michelou Is this one ready to go?

philwalk commented:
@michelou I'm working on tests to thoroughly verify this PR, but they aren't ready yet.
I have done lots of manual testing, so I believe it's safe to merge.

philwalk commented:
- added tests to verify windows wildcard classpath in [dist/bin/scala
- added ability for tests to verify [dist/bin/scalac] generated java command line

philwalk commented:
@michelou - 

> @philwalk Your addition on [lines 122-126](https://github.com/lampepfl/dotty/blob/8c72495dc7e5008e31e0fdeb8f1a64654cdf4035/dist/bin/scalac#L122) in file `bin/scalac` is temporary right ?!

It can be temporary if there are objections to it.   We possibly can generate an edited copy of `scalac` and call that at test time.

The reason I added it was to be able to test `scala`, `scalac` and `scaladoc` scripts.   I originally tried to diagnose the same thing from inside of a running script, but there are challenges and I haven't got that approach to work yet.    I originally got 

In the test file [ClasspathTest.scala](https://github.com/lampepfl/dotty/blob/8c72495dc7e5008e31e0fdeb8f1a64654cdf4035/compiler/test/dotty/tools/io/ClasspathTest.scala) on line 38:
```scala
      val cmd = Array(bashExe,"-c",s"SCALAC_ECHO_TEST=1 dist/target/pack/bin/scala -classpath '*/lib' $relpath")
```
To see what happens, you can remove the SCALAC_ECHO_TEST from the test:
```scala
      val cmd = Array(bashExe,"-c",s"dist/target/pack/bin/scala -classpath '*/lib' $relpath")
```
It fails in Windows with a complaint about mismatched class file versions.

philwalk commented:
When I run sbt tests from a Windows command shell, the new classpath test fails, so I'm working on how to handle it.

michelou commented:
@philwalk Note about style : please add a space after commas, e.g.
```val cmd = Array(bashExe, "-c", bashCmdline)```
 instead of
```val cmd = Array(bashExe,"-c",bashCmdline)```

philwalk commented:
Classpath tests now cover the two most difficult use cases:
A classpath consisting of a single wildcard entry, specified on the command line:
```bash
$ scala -classpath 'lib/*' script.sc
```
A classpath consisting of a single wildcard entry, specified in a hashbang:
```scala
#!/opt/scala3/bin/scala -classpath 'lib/*'
def main(args: Array[String]): Unit =
  val psep = sys.props("path.separator")
  for entry <- sys.props("java.class.path").split(psep) do
    printf("%s\n",entry)
```

Unfortunately, I checked in a bug in [dist/bin/scala], I'll push the fix after the current checks complete.
BTW, is there a way to cancel the current checks to avoid wasting resources?

philwalk commented:
@michelou 
> @philwalk Note about style : please add a space after commas, e.g.
> `val cmd = Array(bashExe, "-c", bashCmdline)`
> instead of
> `val cmd = Array(bashExe,"-c",bashCmdline)`

I'll go through and check for style errors.

philwalk commented:
I'm ready to push a bug fix and corrected style errors, should I wait until the current checks are completed?
It would be nice if we could cancel them.

philwalk commented:
The comment on the most recent commit should have been:
'added missing shift in new -classpath* case in dist/bin/scala; fixed various style errors'

michelou commented:
> The comment on the most recent commit should have been:
> 'added missing shift in new -classpath* case in dist/bin/scala; fixed various style errors'

Strange. Job [0dc35d8](https://github.com/lampepfl/dotty/pull/11633/commits/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344) seems to be still running ?!

philwalk commented:
> > The comment on the most recent commit should have been:
> > 'added missing shift in new -classpath* case in dist/bin/scala; fixed various style errors'
> 
> Strange. Job [0dc35d8](https://github.com/lampepfl/dotty/pull/11633/commits/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344) seems to be still running ?!

@michelou It seems to be finished now (4 hours after your comment).

philwalk commented:
close and re-open to trigger a rebuild ... there does seem to be something odd about the community build status reporting.

philwalk commented:
> @philwalk Your addition on [lines 122-126](https://github.com/lampepfl/dotty/blob/8c72495dc7e5008e31e0fdeb8f1a64654cdf4035/dist/bin/scalac#L122) in file `bin/scalac` is temporary right ?!

@michelou I'm ready to push an updated version of [compiler/test/dotty/tools/scripting/ClasspathTests.scala](https://github.com/lampepfl/dotty/blob/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344/compiler/test/dotty/tools/scripting/ClasspathTests.scala) that eliminates the `SCALAC_ECHO_TEST` change to [dist/bin/scalac](https://github.com/lampepfl/dotty/blob/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344/dist/bin/scalac),  based on testable modified copies of [dist/bin/scala] and [dist/bin/scalac].

Are you aware of any other changes that are needed for this PR?

michelou commented:
@philwalk Your file [`dist/bin/scalac`](https://github.com/lampepfl/dotty/blob/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344/dist/bin/scalac) differs from current head (e.g. lines 92-93). Please check it again.

philwalk commented:
> @philwalk Your file [`dist/bin/scalac`](https://github.com/lampepfl/dotty/blob/0dc35d84c0fcc398f8e20e19d1ed9b410ac9d344/dist/bin/scalac) differs from current head (e.g. lines 92-93). Please check it again.

@michelou - I will push my changes now, the link above is to the original proposal which has the objectionable lines.

liufengyun reviewed (approved):
LGTM, thanks @philwalk 🎉

liufengyun commented:
@michelou I'm merging this now. Feel free to open a new issue if you spot some problems. Thank you.
